packages/cosmic-swingset/test/export-storage.test.ts

```diff
@@ -1,29 +1,44 @@
-// @ts-check
-import test from 'ava';
+import type { TestFn } from 'ava';
+import anyTest from 'ava';
 import { exportStorage } from '../src/export-storage.js';
 
-const makeBatchChainStorage = published => {
-  const deleted = [];
-  const batchChainStorage = (method, args) => {
-    const lookup = path => {
-      let node = published;
-      if (path === 'published') return node;
+const test = anyTest as TestFn;
+
+type PublishedNode = {
+  _?: unknown;
+  [key: string]: PublishedNode | unknown;
+};
+
+type BatchChainStorage = (
+  method: string,
+  args: any[],
+) => string | string[] | [string, string][] | null | void;
+
+const makeBatchChainStorage = (published: PublishedNode) => {
+  const deleted: string[] = [];
+  // @ts-expect-error XXX return type
+  const batchChainStorage: BatchChainStorage = (method, args) => {
+    const lookup = (path: string): PublishedNode | null => {
+      let node: PublishedNode | unknown = published;
+      if (path === 'published') return node as PublishedNode;
```

Comment:
This function can be simplified.  (I wish I could do it in one suggestion, but Github doesn't let me cross the deleted lines).

```suggestion
      let node = published;
      if (path === 'published') return node;
```

Reply:
I tried that and hit some problems. I'm going to leave it.
<img width="622" alt="Screenshot 2025-03-27 at 9 30 37 AM" src="https://github.com/user-attachments/assets/26038a1e-6427-40ff-bcad-fda1bf37ed98" />

Reply:
@michaelfig is right, this can be simplified—the only noise I couldn't quite eliminate was `args as $type` casting in `batchChainStorage`: [TypeScript Playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAggygYQJJIDIHsDuEBOUC8AUFFAD5QDkAhhcWZQEa0nkUDGz9FAJp6xH0oAzQRQDmogBaiAlqIBWogNaiANqIC2ogHaj0osKICOonKIDOo4KICuogG6jMogB6iQogF4UA3IVCQsIgoAKpgkHj4UGERbFTmEAA88MhoWLgAfH4B0CkoqBDAwLgEQakY2HjkeUgxuH5s6NrmwFAAKhAtACJUwFQA0hAgpRQA+r7+4NAACjYMqjLmkhDcAHLo3NBRAN5QANod3b0DQwC6APwAXFA22kraWNpQAL5QAGRQ23R7-VAyTwADAAk2xqBSKuGeIJaOH+YmeANO11m80WyzWGwgfmefkIAHo8VAkFBuE0KK17lgoJIqcB0FAhDZVKphjkoACABSrKgaCAAGlgODE5gAlAQMlAAEqFGw4ZoAkmLMC9NiSAUMGytYCSRZQMSFcxQNiqdAJAB0k0CACEVZIEJIqP84HScFR9TBpsSdnR9cBrhzldrrjC4WL8BKQ9oxPRtEzVH4SBBtMBYZ1-YHJMGU6HxYLXSBEnslEMs7CowL7FRVDYIFcoJGxKcsnRVTJVNwcEn071M-Xs1GwxH+2I9qcE-XCgB1GTa9Ca9bAGRCED+s1rpPZtN5qgFvYZ0twitVmt1htNwdQezoGTcbHZKZQG3AVX2x3aZ3oV36xJ0ACyUAgFxim0bhDWLEB0CER9bVfJ0XTdCAPS9KBwMg6DnztB04M-BCkL5QgJSiDk6F5WduGuX98JIKghXMZEaJ5QpcHMRInxfLD33g91PT2X8m3wi9pWAWVtDaKZWJgjiPy-RCeL45tCEaZpWg0KhizYzC32khDSgDOYFiWFZkX0tEVnWTYLy+EglJaEkIFUJjyL7MsR1OUpR3HGzWhNdAlBsMBdP3ZzQ2M1FDIxTYYzjXMrJIKAHNaB5IqiMATPC8cSCXKAAx7Ah8CiChUrC9EKDFTthLlKAkqxOgSCET9sq8lChigNCMzNTswFUKg2AgDk8QAPSKgz0QAHTNPEBQoUqzXMLqZw5CgzVKsVYriv4oI5cC8qiQ5gB6PpBhAVbavWqBtRwKkAFEcEunAOWBbY9oOk4QFeXVOwSHB7BWBkGuKFo4QBEUMri55TsyzacjQ6qoAAQnyyh0AYeQIDYawyHIWH8qiWNmTKmVKrx+MIaqzFSmqosWviKAqG0EAx1JrKOTh6qCYqp5idB8H1vKkSyc2cccUIFsmlsjdU0NIigobEVripldgvLS9j1rA8o1OUcYtOprsfi9BfP8nLtRB07mdZzF2f5jydbF1o5z9fYGwFBtzk1tyoht9amtbdsXtKAB5FG0eAM0JZkToOTZ2nDRgW6d0LZ2oBREazMxJtQfqvAOSahWBV97g3LQguXpOs6Nuy7acfaTp9uOI6xSUxdYxq8vmYLs1RnhxHbk2IR-hWMvy9azUzVSpYOTz9kQQ70YEVOU3y9eeyEk+UmSAdsebAnhWF9BkgebB06+cqh2hdxay7agBhJK0riIGuDTYM4nD9V00iaW4AUaOFSzTvMTAM5VTZQ-hsIe3t4jQHEIUCg1w1pnVzhmD2tNaIxydsORmw8jRXz1j5PyYBjaSEXsPZm2NEbEytkTOM+84qkMxJ3HaNwQIQH7toQeUAT6c2oevDhhNAQgmqp3BENDD4QNXhQcOnRYFrywYgnsyCf6GhpnsM8NCSCcIAsmSWhDiHrVEXFOI4iC6dl0HAnhcjtQKNQco1RPCNGSPMDos0qkCGT3AgvXM4FdFH3LoYqBCRgDTlnPOdAi5lzSPgetLOjUr57nka1KCii0EqIwaOcBw9NgJRWFvCeGZvF6LsXwkRpM+5UCZH6HhF1rq3U-A9B4rQZAaC6tcEEoDuAIm8YfYW6i+GfGvrfbCMkBSZMci8O8IsmqOzpsMHYwt8R4gBsARagEwCfmANpN+gEeRdQEAKVo4Y16xSaistZGyIBwDmCmCAW4Gxa09oVNK6IzRsDNPoTB2DlIARcJAdGRl0EuRdsON2dz9inT2A84q2SXn6CmnSQwpwqJxXBcNUy3BnmvLNHIBFYKIWpzRdCzFZpdBTQACwACYKDYvWsix5UKMXOCmhoA2lLEWM1Fp83YN8MLPzOS8Uoql1IDJfjJDk8C2BmPLugCVw9RjXHJOgQwiK27SqwdoOBUBZVQHJS8JVZ1ni6vWpgdVmqKBMv0Dq0m+rTpWoPsQpqPVhJVlKCcnA6z74ci5exO+r9+RfNWa6s5FyGBXM6MQ0OmwIBgCukYGwVYOQOtjaoAUKyQ6D2xFAAkpt5mLOWd80598jQORooaNgOAbBCFAO0dAIAQhwC6MnOAv4KB7O1kcq+Sl+5iBVYWiANFA2XM7Dc4cXs4ourdT6oNIa6JK1ckLUoYqi04H7cGwd06aWQt4FS0deaA330nau+WpNcWorHuYDQZokBwDaM2o9KLwpmkwFWBKN7y7HvvWAS69hFgyCaNMA2qhnGFFhGwSwuqt1zMvp8u96JfxvlCni8yWwZFxTdJ+GQbBuS8mnZE6+roQLGvluCy916BScvQDRbgSAuhyoYORnA3AAAMABGMlABmElFAXinDcjasGuqwBnu7SQYjQm4ptGrbW6jyGsGkWA9hjV6qljoHOTITwD8oCsb5bxpeBqLV6N1Y+5khRROoeAyx1jBHKBAms5x7Tem4ofvQF+8wP7tB-vQKoUTsn0Pyc1bsOkfRVC-n+MUbgCAmg-VdX8slDGtO6t4xBj5tlOVCrOcM+yozXhRAFRAJ+Ul3XQZWLB-4dqr4JqdVEMdZziLrU9ZpQZCFdUdpkGIMOO7x0yX3dc8wTWmidueYu5dU7EVhrNBGqNMa43laTaC6lb6nkCfPcRllOLCtosWxeq9ZpxM1rrStuba3T1La2ztyTgHswgebZQPYyNUboygEHO7wBTj7aRfN7Jhnn1bpIOuvFD6n2FDNKZ9D5mrsUGs0CV7P33vrc-d+39-7ztybBzd4O93Hshxe1uveIsSChxTDWDkIzQuYu0MaGwmxHEw6O5tto22JN7ZFKbV4mbcSMjJ4uJofr80+o5GuIHtFri3EpJgbQaTrjTM+ODIAA)
```ts
  const lookup = (path: string): PublishedNode | null => {
    let node = published;
    if (path === 'published') return node;
    for (const key of path.replace(/^published\./, '').split('.')) {
      if (key === TestDataKey) {
        throw Error(`${TestDataKey} is reserved for testing`);
      }
      if (typeof node !== 'object' || node === null) return null;
      node = node[key as any];
      if (!node) return null;
    }
    return node;
  };

  const entries = (path: string): [key: string, value?: string][] => {
    const node = lookup(path);
    if (!node) return [];
    const out: [string, string?][] = [];
    const childData = Object.entries(node) as Array<[string, PublishedNode]>;
    for (const [key, child] of childData) {
      if (key === TestDataKey) continue;
      if (child._ !== undefined) {
        out.push([key, `${child._}`]);
      } else {
        out.push([key]);
      }
    }
    return out;
  };
```

Reply:
After fighting with Typescript for too long to simplify the code, I have a different suggestion: use a symbol for the TestDataKey, and all sorts of typing and namespace collision problems vanish.

[TypeScriptPlayground](https://www.typescriptlang.org/play/?#code/MYewdgzgLgBA+jAvDAygTwLYCMQBsAUA5ACoCm0AIgIZRUDSpahAlANwBQUaADqTAAoBXLLgCWEABakAJgDkQ0vonYwYAbxWqYAbQDWALhjQATqLABzALqGhI8VLkLSHLTriWA-IcFhdYEADuYC4wAL7s7AD0kTAAkjDS4ISwfoEwEmlQIDAAZoK4uGgwXLwwAAb4slQYpAA0MACCxuYQzEgAfDAASqRQgsaQZQni3DTAEvVYgrBQEuIw5r0QMMC4IBCkAHScPHwAQmMSAMISVGYoWcZUiw388cgaqotQhvijs4YmZuZtiJ1fFhgAB8YGB8rgQqQwFBTORXu8JJ8Yd9fp0mlc0AAePSMJGmCz1ABuVFwglIXiMyIslnaIXGolw0mMUPhNERlPxPw6HO+2ksIQ2UAA6qJZiBpvIoKIcmhXpt5VDkXDGsYMdiEXjvkSSWSKQCrO1UTBCSBRNIOKEOCV9ocTmcwBcQFdFpjNABZGCkAAeUCh0mWukYIByMAOUHGdvOl2upFu9xggbQwdDttOUadMbjtXYnWQ+E0NTF0kMbuzqiozQgNgr1V6pGMEExYYjaYd0Zud20bpp2aNPT6A2Iuybqftjudsc73dpEVAkFgGCogebx1b45jSBgb2EYkkMhsO-sMnkiiNjxW4GgCVIuDrxZ5Fj5m75dMvsDWIF0gm4m7ebM1FjMAedh7o4ijAqC4Lcueqi3rA-jgcg3CHqBISqNKW4IkgiDIIQyEgQ4LAwMyA5gKCThobkTpbnOV6JjAyYIpszLcLgVDAKQ+CRAAevhu4OAAOpskT1IQLCbBArGikQmwsG0MFaBh+DWsmCF8AAhDhMCECAWAAFakMAUCEMCIJqdhyBggUbQkf0ZFWRCmiuOZllODiaD8k5ikhvg6lqTZvR2ZBBSUeEri2QM5GKCEloRKotGwIqsLLHmGoPj8hjuQB5jaqS5LZZYT5-OoXkJVFSgwB+X7cH+sxsF5Sl+U4AWkTonmuGV4ovDo+r1PqHiFZYz7tVoZX0oy1C0JuADy+mGVAmxJaI5D4P5MBUMs6JUFi2i9QIKEOCepA0pROTUfgZXufU43SENyY3ZNVDyV56E+TdmwIJpyA+IoORmDIz2uK4XWbMhkj4Fd5QACRqDd2juKEZSWPVQNhJ6uAbCVqOqCDYMSBDibI5RqhhVopOqBFZFdTFHCaGVWCjumE6GCukZthmiy-oWGTSPUFYtGeXkQAEorjFu3MKIDQPABtfCEM8hCGApHVvjoCJDcg-PLBtPVUlYxOjarLmVSAn7frVEgo9jSkuVpDktUFDkG69W5qR9FkwD9pB-WAAPEYFkVOy9WiU9DahqfDliIwb5OjbL2lLeQitY9jl3q5uWvrcsu16yNqOh4nEAW1brix-F8eEDdzJgMnyuG-Oatshr62VlnuucnnQMF9CyXF5si41QTjDI9yiYl2Twcy5jhCCiKYoSiAUoyrXweqGdxg0ar2jp8mmc6znHd8lL2PXnBMig4I4MIuPper-7pEx8Hv1UPkLx37MxhpAAoqqTr4GU-hYCiAwKxQwMMJbSERuPUmsUKYBzImoGADNwyrjHO2Oop87xhAtLTMq3UqBgCKA8WKURIi+mgEQb03AnRQHXJzb01RWKkEIPUWAxUNDnjKlQmhdDSAoGEDCUgyp9RFR0HhA659gCbBAIQEaXCvS8CMvuduWp0oDVEdoLy2hxEEUkdIlh2ksjcFkWWLQ2i+JHmkJsKRIBNiiBMVonR-E9G2NEJsGuokAAsAAmBxrhzESKsTYzYAQDGEAwKbPxMBPJ01Vog5BLY0Ecz4KETci5lyM3ZhOfAClgBK2DiAfJJ9I6GGSCAYxpjUaiCKSfUESs3DWBgD4sIlSgahFaa4AI9SSnaQiTIlpwd2leSGSTK2ZV2J9BJJubhxhaHoPwAk1BTMYz1BmXM5J-CsCCPIFbBaihSDcC-gAR0ECSfAEzTm4FWQo+aAMLQwGiPVUh5CoCUIUTw9BKxbwVmWMAYwggchcBgMQEAaAACqKAKACBQG6AxbDOgcNiQ3Ocf1zA1Pit84wvDNnbKrOlF8Xk1nYoEcyYResCUk03LkzFxKtmkrxQE3R0gomqCJegnF9LMrBycZY0GEAMCbFiCgYgLDuUWNAiEkkcFRWox5RK7gn9CTiFEOAfgptcD916KYYAEAZVaEsKYkh8VVbiocG6e0wFnFgQqgpa4TpRDACqDUPFdcsBXDAPeRBPTtDaKFSK+o8SQAVmkLECgpScDBoAAwAEZvEAGZPEmVCJYIaIyyatO4Py9FWg-XZtcCC8FkK81A0LNql1DT6mSBAHw0QAAvUghg41ozTajFt6bhmtICFK3oxa7XatjXG7p7hSlQ1HUm1pbaFUgCVRAFVYA1V4GLaWh15aemIKyLQXAbozC+mkEccAhJ6y7sMN4yNzaJ2Gtpsahu8TMm8PqIoM+0g0bIHSaQVma55mmpkOaswYzVYXKmcgNlyT8yuEWWze9pVwCosWu82ZtLcWtJRaIcw1iaXspJUIiApjdmbH2Yck5ZzANXJ0I479VjM0Cr9SysRFG+XUeFZsAtEKKC0cZVahjgqmMschZq5EOqwnaF0gZIyMBZqiagJYdjcqHCSoKL0GT9Gu0KYWn2h1A6wmjqhkpwJoNFXKtVeq-jZahMifmuJuaRlpMGs0ETOKxRNgwjJPgR9d47FgFWIIRQRdZPnyo9x4gzHQWsbkvcx5tM8iealOAT08H1nZPlJsLW3hfD+CCEfQwBCihqHCEAA)

Reply:
I didn't mean to nerd snipe you guys. Do you think all this change is worth it to avoid the casts?

Reply:
It was worth it to me to improve my understanding of index and mapped types.  If you don't care to make this change for the current PR, I may propose it later when I'm cleaning out my inbox.

IMO, making the data structures coherently typed is worthwhile, and as a knock-on benefit, it avoids the casts.

Reply:
My experience is similar. I was surprised to not find a way to apply less narrow constraints to particular keys in a TypeScript Record type, but came away with better understanding of it. Refactoring is worthwhile IMO, but need not be considered blocking for this PR.